### DISCUSSION TITLE
Example of heightfield map?

### DISCUSSION THREAD

**Discussion by patrikpatrik** [DanielChappuis/reactphysics3d#272] [Q&A] [answered]
Can you give an example of a heightfield map? I'm starting with a simple 64 x 64 black and white heightmap *.png file and my cube appears to be hitting something, but it's not lining up to the texture my heightmap is mapped to?

I'm using `stbi` library and use this to generate a heightmap terrain:

```
// terrain size & count
const int HEIGHTMAP = 64;
float heightmapArraySize[HEIGHTMAP][HEIGHTMAP]; // <-- for heightmap reactphytsics3d

// height map (for terrain generator)
	int heightX, heightY;
	unsigned char* heightMapData = stbi_load("./heightmap64.png", &heightX, &heightY, NULL, 1);
// storing into array
	for (int i = 0; i < HEIGHTMAP; i++)
	{
		for (int j = 0; j < HEIGHTMAP; j++)
		{
                        //heightmapArraySize[j][i] = heightMapData[HEIGHTMAP * j + i];
			heightmapArraySize[j][i] = getHeight(j, i, heightMapData, HEIGHTMAP);
		}
	}
```

I suppose my question would be, what values are supposed to be stored in the heightmapArraySize array, is it the actual heightmap int greyscale values from 0 - 256 or is it float type numbers? Here is the heightfield example:

```
// example of how to create heightFieldShape
	const int nbRows = 64;
	const int nbColumns = 64;
	float minHeight = -40; // <-- some value im not sure of, setting this min to 0.0f throws an assertion error @ 440 heightFieldShape.cpp
	float maxHeight =  40; // <-- some value im not sure of

	// create a heightfield map
	rp3d::Vector3 heightfieldMapPos(0.0f, 0.0f, 0.0f);
	rp3d::Quaternion heightfieldOrientation = rp3d::Quaternion::identity();
	rp3d::Transform heightfieldTransform(heightfieldMapPos, heightfieldOrientation);
	rp3d::RigidBody* heightfieldBody = world->createRigidBody(heightfieldTransform);
	heightfieldBody->setType(rp3d::BodyType::STATIC); // <-- set heightmap to static

	// Create the heightfield collision shape 
	// last param is done like rp3d::HeightFieldShape::HeightDataType::HEIGHT_INT_TYPE
	rp3d::HeightFieldShape* heightFieldShape = physicsCommon.createHeightFieldShape(nbColumns, nbRows, 
		minHeight, maxHeight, 
		heightmapArraySize, rp3d::HeightFieldShape::HeightDataType::HEIGHT_FLOAT_TYPE);
	// add collider
	rp3d::Collider* heightmapCollider;
	heightmapCollider = heightfieldBody->addCollider(heightFieldShape, rp3d::Transform::identity());

```
I'm not quite sure if I'm even doing this correctly, thank you!

**Comment by patrikpatrik** [comment, marked as answer]
Ugh... I always do this but I think I answered my own question. The array is wrong, should be:
```
// terrain size & count
const int HEIGHTMAP = 64;
float heightValues[HEIGHTMAP * HEIGHTMAP]; // <-- change here
// height map (for terrain generator)
	int heightX, heightY;
	unsigned char* heightMapData = stbi_load("./heightmap64.png", &heightX, &heightY, NULL, 1);
// storing into array
for (int i = 0; i < HEIGHTMAP; i++)
	{
		for (int j = 0; j < HEIGHTMAP; j++)
		{
			int arrayIndex = j * HEIGHTMAP + i;
			heightValues[arrayIndex] = getHeightFieldMapValues(i, j, heightMapData, HEIGHTMAP);
			if (i == 0 && j == 0)
			{
				minHeight = heightValues[arrayIndex];
				maxHeight = heightValues[arrayIndex];
			}
			if (heightValues[arrayIndex] < minHeight) {
				minHeight = heightValues[arrayIndex];
			}
			if (heightValues[arrayIndex] > maxHeight) {
				maxHeight = heightValues[arrayIndex];
			}
		}
	}
```
As per HeightField.cpp @ line 171. But most importantly, the terrain y position was always off by some factor. In the `UserManual` 11.1.6 HeightField Shapes, it states:
`Therefore, if you create a HeightFieldShape with a minimum height of 100 and a maximum height of 500, the maximum coordinates of the shape on the Y axis will be 200 and the minimum coordinates will be -200.`
In other words  `terrainPosition.y - (maxHeight + minHeight)/2` for the collider to line up with the terrain heightmap. And now I can change to different heightMaps *.png and objects falling correctly hit the terrain generated, at least for 64 x 64!👍
